docs: add copy button for the site

gweesin opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

closed #18769 

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[X] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

add a copy button for each code blocks

![image](https://github.com/user-attachments/assets/91af90ec-0548-4ea3-b93c-811a29815dc4)
![image](https://github.com/user-attachments/assets/ceb563c8-39db-4049-9393-0e085e5f0f33)


#### Is there anything you'd like reviewers to focus on?

can we make the style better match the current website theme?


<!-- markdownlint-disable-file MD004 -->

Tanujkanti4441 commented:
Hi @gweesin, thanks for PR, but right now CI is failing, could you fix that?

gweesin commented:
> Hi @gweesin, thanks for PR, but right now CI is failing, could you fix that?

sure, it seems this is a lint problem, but now I can't run the trunk command. I will research it later.

gweesin commented:
anybody can help?

```shell
PS D:\opensource\eslint> trunk
node:internal/deps/undici/undici:13185
      Error.captureStackTrace(err);
            ^

TypeError: fetch failed
    at node:internal/deps/undici/undici:13185:13
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5) {
  [cause]: Error: read ECONNRESET
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:218:20) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
  }
}

Node.js v22.8.0
```

Jay-Karia commented:
> anybody can help?
> 
> ```shell
> PS D:\opensource\eslint> trunk
> node:internal/deps/undici/undici:13185
>       Error.captureStackTrace(err);
>             ^
> 
> TypeError: fetch failed
>     at node:internal/deps/undici/undici:13185:13
>     at process.processTicksAndRejections (node:internal/process/task_queues:105:5) {
>   [cause]: Error: read ECONNRESET
>       at TLSWrap.onStreamRead (node:internal/stream_base_commons:218:20) {
>     errno: -4077,
>     code: 'ECONNRESET',
>     syscall: 'read'
>   }
> }
> 
> Node.js v22.8.0
> ```

Seems like a network issue

Tanujkanti4441 commented:
> docs/src/assets/images/icons/copied.svg
     `fmt  Unoptimized svg, fix by running 'trunk fmt'`

i guess recommended command is `trunk fmt` to optimize the SVG that is creating the issue.

gweesin commented:
> > docs/src/assets/images/icons/copied.svg
> > `fmt  Unoptimized svg, fix by running 'trunk fmt'`
> 
> i guess recommended command is `trunk fmt` to optimize the SVG that is creating the issue.

Yes, the error caused by fetch https://trunk.io/releases/1.22.3/trunk-1.22.3.mingw.tar.gz

![image](https://github.com/user-attachments/assets/f7361e1f-126a-4d86-bb07-23666aa16ab9)

but my network really connect it successfully: ![image](https://github.com/user-attachments/assets/4682aa7c-2d9d-4518-bbbf-03391133e38c)

nzakas commented:
@gweesin what operating system are you on?

@det can you help us debug this trunk problem?

nzakas commented:
Thanks for taking a look at this.

All of the code samples now have the  language in the upper-right corner. I don't think we want that.
![Screenshot 2024-09-06 at 11-49-05 Getting Started with ESLint - ESLint - Pluggable JavaScript Linter](https://github.com/user-attachments/assets/00fbca54-d3bf-4c7d-ab79-e3a0e41db7c8)

I think the copy button should be visible all the time, not just on hover. Also, the image disappears completely in light mode.
![Screenshot 2024-09-06 at 11-49-36 Getting Started with ESLint - ESLint - Pluggable JavaScript Linter](https://github.com/user-attachments/assets/45da0329-e598-4fde-98f1-ec26fba1fdef)

Also, we should not be applying this to valid/invalid code in the rules docs as it overlaps.

![Screenshot 2024-09-06 at 11-51-02 array-callback-return - ESLint - Pluggable JavaScript Linter](https://github.com/user-attachments/assets/398ee675-e85b-43f9-a649-7f5fcefccbed)
![Screenshot 2024-09-06 at 11-51-31 array-callback-return - ESLint - Pluggable JavaScript Linter](https://github.com/user-attachments/assets/12d23fbe-d643-4db2-ae0e-134f77222d3e)

gweesin commented:
> @gweesin what operating system are you on?
> 
> @det can you help us debug this trunk problem?

my OS is Windows 11. I can provide more informations if you need.

gweesin commented:
@nzakas Thanks for your review.
the language corner forked from vitepress, I haven't removed it before because it worked properly. Also I can remove it latter.

button's icon disapeers in light mode seems a bug.

shouldn't display in code block for validation demo is the good user experience.

I will fix for all of advertisements. Thanks a lot.

EliSchleifer commented:
@det is on the case - working with reporter to understand why `trunk fmt` won't run on box

Tanujkanti4441 commented:
Hi @gweesin, there are some suggestions, you can check them out :).

amareshsm commented:
@gweesin Just a gentle reminder: there’s some feedback that I'd appreciate you taking a look at. If you're busy, please let us know, and I can handle it from here.

Additionally, I’ve addressed the lint issue for now and pushed the changes from my end.

amareshsm commented:
> @gweesin Just a gentle reminder: there’s some feedback that I'd appreciate you taking a look at. If you're busy, please let us know, and I can handle it from here.
> 
> Additionally, I’ve addressed the lint issue for now and pushed the changes from my end.

Since it has been pending for a long time I will take over the pr complete the remaining changes.

gweesin commented:
> > @gweesin Just a gentle reminder: there’s some feedback that I'd appreciate you taking a look at. If you're busy, please let us know, and I can handle it from here.
> > Additionally, I’ve addressed the lint issue for now and pushed the changes from my end.
> 
> Since it has been pending for a long time I will take over the pr complete the remaining changes.

Thanks! I'm sorry I have no time to fix it recently. Also puzzled by running trunk command.

nzakas commented:
@gweesin has someone reached out to you to discuss your issues with Trunk?

gweesin commented:
> @gweesin has someone reached out to you to discuss your issues with Trunk?

Chris Clearwater told me upgrade the trunk launcher package to version 1.3.2 on Slack, but nothing changed. And then the communication was stopped.

nzakas commented:
Okay, I'll reach out to them and see if we can get you some help.

gweesin commented:
> Okay, I'll reach out to them and see if we can get you some help.

I'm sorry for resolving this issue so late. Here is my solution for the trunk installation:

1. Download `trunk-1.22.3.mingw.tar.gz` using Chrome.
2. Start a local static server to serve the file via an HTTP request.
3. Modify the download URL in `/node_modules/@trunkio/launcher/src/download.js`.
4. The program should now be able to download the file successfully.

It would be better if Trunk could support reading the download mirror from environment variables.

I will continue working on this PR.

nzakas commented:
@gweesin thanks for the update, and glad you got things working. Do you want to finish up this PR?

gweesin commented:
> @gweesin thanks for the update, and glad you got things working. Do you want to finish up this PR?

I have rebased the branch and fixed the bugs you mentioned. However, I couldn't reproduce this particular issue. It seems to be a temporary compilation state during development.

![Uploading image.png…](https://private-user-images.githubusercontent.com/38546/365229902-45da0329-e598-4fde-98f1-ec26fba1fdef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNDAxODEsIm5iZiI6MTczMzIzOTg4MSwicGF0aCI6Ii8zODU0Ni8zNjUyMjk5MDItNDVkYTAzMjktZTU5OC00ZmRlLTk4ZjEtZWMyNmZiYTFmZGVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDE1MzEyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhMmNiMmZhMmJmN2Y4NTFiODIzZmJkODhkZjA5NTI2NDM2Yjk0N2Y0YjNjMDVkOTYxMmIwYjc5MWI5NDE0NGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YVAdj1KiZi97Nzb8AO4WjcPDu_yrHdDOYvISEf4y0qs)

Thank you for your review! We're now ready to revisit the code for further review.

nzakas commented:
I can see the icon now! :tada:

I think in dark mode the button should be darker. Basically, just slightly lighter than the background.

![Screenshot 2024-12-09 at 11-32-42 Getting Started with ESLint - ESLint - Pluggable JavaScript Linter](https://github.com/user-attachments/assets/fd245fb4-0ef8-4642-92f3-95198420fad8)

In light mode, we already have the button as slightly darker than the background.

![Screenshot 2024-12-09 at 11-33-25 Getting Started with ESLint - ESLint - Pluggable JavaScript Linter](https://github.com/user-attachments/assets/ea1604d4-5866-418e-81f9-73b8b23ef9a4)

nzakas commented on docs/src/assets/scss/components/copy-button.scss:
```diff
@@ -0,0 +1,48 @@
+
+[class*="language-"] > .c-btn--copy:hover,
+[class*="language-"] > .c-btn--copy.copied {
+    opacity: 0.8;
```
The semi-transparency is a problem when the button overlaps text, making it harder to see the button on hover.

What do you think about doing a box-shadow on hover instead?

nzakas commented:
At least on my screen, I don't see anything in dark mode and light mode is also difficult to see.

Maybe we should just go with a glow effect using box-shadow? So something along the lines of `0 0 5px rgba(0, 0, 0, 0.4)` (although that only looks good in light mode and can't be seen in dark mode.)

JoshuaKGoldberg commented:
:wave: Temporary notice, I'm about to merge [#19355 chore: formatted files with Prettier via trunk fmt](https://github.com/eslint/eslint/pull/19355). We should hold off on merging other PRs till it's done. I'll post back soon to confirm we're clear. Cheers!
(context: https://github.com/eslint/eslint/pull/19355#issuecomment-2733445086)

mdjermanovic commented:
Seems like this has already been implemented by https://github.com/eslint/eslint/pull/19512, so we can close this PR?
